### PR TITLE
[IOSP-437] Fix logic when creating MergeService (leading to auto-accepting PRs for new target branches)

### PR DESCRIPTION
**Pull request by AliSoftware**
Ticket: https://babylonpartners.atlassian.net/browse/IOSP-437

After deploying #39 recently, we realised that there was a flaw in the logic: every time a new Pull Request targeting a branch that doesn't exist yet is created, leading to the creation on a new MergeService, we forgot to filter out those PullRequest events sent by GitHub to only create the new MergeService if that PR was supposed to be included in the first place)

Consequently, any new PR targeting a new target branch was automatically triggering the creation of a new `MergeService` and add that PR to the `MergeService`'s queue… even if that PR didn't have the integration label and wasn't ready to be merged (basically as soon as the PR is created)

This PR fixes that flawed logic.

### CODE REVIEW

File: Sources/Bot/Services/MergeService.swift
Comment (dmcrodrigues) [inline Sources/Bot/Services/MergeService.swift]: Do we really need a type for this? Why not using free/static functions? I don't find it a type particularly discoverable 😕

Comment (dmcrodrigues) [reply Sources/Bot/Services/MergeService.swift]: When I say a type I mean using failable initialisers.

Comment (AliSoftware) [reply Sources/Bot/Services/MergeService.swift]: tbh I initially created a `static` free function at first, but the signature felt a bit strange to read, especially since I named that `static func outcome(for: …) -> MergeService.Event.Outcome?`… so since the func was named `outcome` and returned the (already existing †) `Outcome` type, I figured it would make more sense to make it an initialiser on the existing type

_(†) The `Outcome` type is not new and used in the `Feedbacks` of the `MergeService` machine already, all I did is move the logic from the `Feedback` to a `init?` to be able to _also_ call it from the `DispatchService` to decide wether to create a new queue or not)_

I'm ok with moving those to free functions if you prefer it to using an `init?` – just know that even if I do that the `MergeService.Event.Outcome` type will still exist as it's used in the state machine already.

File: Sources/Bot/Services/DispatchService.swift
Comment (AliSoftware) [inline Sources/Bot/Services/DispatchService.swift]:
```suggestion
                // Only create a new MergeService if the PullRequest for which we received an event is supposed to be included
```

File: Sources/Bot/Services/MergeService.swift
Comment (dmcrodrigues) [inline Sources/Bot/Services/MergeService.swift]: `map`+`skipNil` can be flatten using `filterMap`.

File: Sources/Bot/Services/MergeService.swift
Comment (dmcrodrigues) [reply Sources/Bot/Services/MergeService.swift]: That's ok let's keep it, is just a detail 👍

File: Sources/Bot/Services/MergeService.swift
Comment (AliSoftware) [reply Sources/Bot/Services/MergeService.swift]: Good point, not sure why we didn't spot it earlier 👍

File: Sources/Bot/Services/DispatchService.swift
Comment (dmcrodrigues) [inline Sources/Bot/Services/DispatchService.swift]: Aren't we leaking implementation details in doing this filtering at this point? Because `MergeService.Event.Outcome` should be something private to `MergeService` and not exposed even internally, same for Events which are now exposed. 

I understand that we may end up creating a new `MergeService` without a need for it, if the PR is not labelled with the integration label, but that's an implementation detail and the `MergeService` will be released eventually so that's not entirely harmful in creating this instance.

Comment (AliSoftware) [reply Sources/Bot/Services/DispatchService.swift]: So you're suggesting we create the MergeService every time a new PR targeting a new branch is created, even if that merge service will directly go from `.starting` to `.idle` and be deleted right away?

I'm ok with that, just wondering if it's worth it compared to refactor that to just expose a `MergeService.shouldInclude(…) -> Bool` func (that would privately check if the `Outcome` is `.included` but will keep `Outcome` private?

Comment (dmcrodrigues) [reply Sources/Bot/Services/DispatchService.swift]: Either we do it for all events or none in my opinion, as a reader/maintainer is very confusing why we only filter part of the events... Because if the service already exists we just forward the change to the service which will eventually filter the event internally in the feedback that I've mentioned.

```swift
existingService?.pullRequestChangesObserver.send(value: (event.pullRequestMetadata, event.action))
```

Conceptually `DispatchService` should only be responsible by keeping the merge services and forward the events accordingly, now it's already doing some logic where it becomes aware of the concept of filtering Pull Requests that are not labelled with a certain integration label which is a concept that's expected to be part of the service itself.

Comment (AliSoftware) [reply Sources/Bot/Services/DispatchService.swift]: Yes I agree that I'm not a fan of that logic of only forwarding the event to the service if it already existed and use the `initialPullRequests` otherwise. We might want to improve that indeed.

I'll try to work on this, allowing the `.starting` state to also be able to listen and react to `pullRequestEvents`. _(Tried that in the original implementation but let to some unexpected complications about tracking state changes during startup sequence 😉 will try to revisit now that the code has settled)_

Comment (dmcrodrigues) [reply Sources/Bot/Services/DispatchService.swift]: > I'll try to work on this, allowing the .starting state to also be able to listen and react to pullRequestEvents.

I'm not entirely following, we are injecting `initialPullRequests` to the `MergeService` as part of the initialisation so that filtering can be done directly in the initialiser or inside the feedback for `starting` status, I'm not sure why we need to listen and react to `pullRequestEvents` 😕

Comment (AliSoftware) [reply Sources/Bot/Services/DispatchService.swift]: ~We can't really do the filtering about the type of PullRequest event in `initialPullRequests` since we'd have the PullRequests directly here, not the event, so we wouldn't be able to check if the event leading to this was a `.opened` or `.labeled` or whatnot~

Wait or were you suggesting to just filter on the `pullRequest.isLabelled(state.integrationlabel)` – instead of filtering on the event? (I thought that at first but was afraid of possible edge cases, but in fact maybe that could be enough indeed 🤔 )

Comment (dmcrodrigues) [reply Sources/Bot/Services/DispatchService.swift]: Events are irrelevant for the initial list of pull requests, the only thing that matters is if the PR has the integration label or not, during the start up of the bot we don't have any events being fired, we just load the entire list of pull requests and start from there based on the presence of the integration label.

Comment (AliSoftware) [reply Sources/Bot/Services/DispatchService.swift]: Gotcha. But that means if we go that route, we'll keep the `existingService?.pullRequestChangesObserver.send` iow the fact that the `DispatchService` knows when to send those events (when MergeService was already existing) vs when not to (when it just got created), and I didn't like that either and wanted to improve this. But maybe I can send the event in all cases and filter in the `Feedback` if it's `.starting` after all, so this should work.

File: Sources/Bot/Services/DispatchService.swift
Comment (AliSoftware) [inline Sources/Bot/Services/DispatchService.swift:95]: Notice that we now send the event in all cases, and let the `MergeService` filter that event out (the `Feedback.pullRequestChanges` of `MergeService` already filters out those events if the service was just started, using `predicate: { $0.status != .starting }`).

Which means we don't leak logic and implementation details about the internals of `MergeService` to `DispatchService` anymore, it doesn't have to know that logic about "don't send event if service was just created", instead we dispatch in all cases and let the `MergeService` do the filtering

File: Sources/Bot/Services/MergeService.swift
Comment (AliSoftware) [inline Sources/Bot/Services/MergeService.swift:46]: This ends up being the main fix for the initial bug

File: Tests/BotTests/MergeService/DispatchServiceTests.swift
Comment (AliSoftware) [inline Tests/BotTests/MergeService/DispatchServiceTests.swift:212]: Notice the new behavior: with the new implementation that avoids `DispatchService` having to know about the internal logic and impl details of `MergeService`, that means that it always creates a `MergeService` even if that `MergeService` ends up deciding there's nothing to do and gets immediately destroyed 🤷‍♂

File: Tests/BotTests/MergeService/MergeServiceTests.swift
Comment (AliSoftware) [inline Tests/BotTests/MergeService/MergeServiceTests.swift:254]: Notice the new behavior: with the new implementation that avoids `DispatchService` having to know about the internal logic and impl details of `MergeService`, that means that it always creates a `MergeService` even if that `MergeService` ends up deciding there's nothing to do and gets immediately destroyed… and gets recreated soon after when we receive the `.labeled` event.

**Review by AliSoftware — COMMENTED** [review]
@dmcrodrigues New commits should address your feedback 👍

File: Sources/Bot/Services/DispatchService.swift
Comment (AliSoftware) [inline Sources/Bot/Services/DispatchService.swift:48]: `DispatchService` shouldn't know about the logic related to deciding which PR to filter in or out, which is why this has been removed.
That filtering will be done by the `MergeService` directly instead, to avoid leaking any implementation details and logic of the `MergeService` to the `DispatchService` (even if that means creating a `MergeService` only to see it immediately go to its `.idle` state and being destroyed)

File: Tests/BotTests/MergeService/DispatchServiceTests.swift
Comment (dmcrodrigues) [reply Tests/BotTests/MergeService/DispatchServiceTests.swift:212]: For now... we shouldn't exclude doing something in the future, imagine if we introduce the ability to monitor the PRs to check the number of reviews pending and notify that the PR is ready to be integrated? e.g. after 2 approvals In that scenario the service may stay alive.

**Review by dmcrodrigues — APPROVED** [review]

**Review by danspndl — APPROVED** [review]
LGTM 🚀

**Review by ilyapuchka — APPROVED** [review]

**Review by inamiy — APPROVED** [review]

**Pull request merged**
